packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+      totalStartingWidth - startInset - sideInset;
+
+  @override
+  _FrameBasedTimelineFlameChartState createState() =>
+      _FrameBasedTimelineFlameChartState();
+}
+
+class _FrameBasedTimelineFlameChartState
+    extends State<_FrameBasedTimelineFlameChart> {
+  static const startingScrollPosition = 0.0;
+  ScrollController _scrollControllerX;
+  ScrollController _scrollControllerY;
+  double scrollOffsetX = startingScrollPosition;
+  double scrollOffsetY = startingScrollPosition;
+
+  final List<FlameChartRow> rows = [];
+
+  TimelineController _controller;
+
+  StreamSubscription _selectedEventSubscription;
+
+  int get gpuSectionStartRow => widget.data.uiEventFlow.depth;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _controller = Controllers.of(context).timeline;
+    _selectedEventSubscription?.cancel();
+    _selectedEventSubscription =
+        _controller.onSelectedTimelineEvent.listen((_) {
+      setState(() {});
+    });
+  }
+
+  @override
+  void didUpdateWidget(_FrameBasedTimelineFlameChart oldWidget) {
+    if (oldWidget.data != widget.data) {
+      setState(() {
+        _scrollControllerX.jumpTo(startingScrollPosition);
+        _scrollControllerY.jumpTo(startingScrollPosition);
+      });
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    // TODO(kenz): improve this so we are not rebuilding on every scroll.
+    _scrollControllerX = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetX = _scrollControllerX.offset;
+        });
+      });
+
+    _scrollControllerY = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetY = _scrollControllerY.offset;
+        });
+      });
+  }
+
+  @override
+  void dispose() {
+    _scrollControllerX.dispose();
+    _scrollControllerY.dispose();
+    _selectedEventSubscription.cancel();
+    // TODO(kenz): dispose [_controller] here.
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Container(
-      color: Colors.black26,
-      child: const Center(
-        child: Text('TODO Flame Chart'),
-      ),
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return Scrollbar(
+          child: SingleChildScrollView(
+            controller: _scrollControllerX,
+            scrollDirection: Axis.horizontal,
+            child: Scrollbar(
+              child: SingleChildScrollView(
+                controller: _scrollControllerY,
+                scrollDirection: Axis.vertical,
+                child: _flameChartBody(constraints),
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  Widget _flameChartBody(BoxConstraints constraints) {
+    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
+    final height = math.max(constraints.maxHeight, widget.height);
+
+    return Stack(
```

Comment:
add a TODO to rewrite this using slivers instead of a stack.

Reply:
done.